sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java

```diff
@@ -400,33 +436,147 @@ private PAssert() {}
 
   ////////////////////////////////////////////////////////////
 
-  private static class PAssertionSite implements Serializable {
+  /**
+   * Output of {@link PAssert}. Passed to a conclude function to act upon.
+   */
+  @DefaultCoder(AvroCoder.class)
+  public static final class SuccessOrFailure {
+
+    private final boolean isSuccess;
+    @org.apache.avro.reflect.Nullable
+    private final PAssertionSite site;
+    @org.apache.avro.reflect.Nullable
     private final String message;
-    private final StackTraceElement[] creationStackTrace;
+
+    private SuccessOrFailure() {
+      this(true, null, null);
+    }
+
+    private SuccessOrFailure(
+        boolean isSuccess,
+        @Nullable PAssertionSite site,
+        @Nullable String message) {
+      this.isSuccess = isSuccess;
+      this.site = site;
+      this.message = message;
+    }
+
+    public boolean isSuccess() {
+      return isSuccess;
+    }
+
+    @Nullable
+    public AssertionError assertionError() {
+      return  site == null ? null : site.wrap(message);
+    }
+
+    public static SuccessOrFailure success() {
+      return new SuccessOrFailure(true, null, null);
+    }
+
+    public static SuccessOrFailure failure(@Nullable PAssertionSite site,
+        @Nullable String message) {
+      return new SuccessOrFailure(false, site, message);
+    }
+
+    public static SuccessOrFailure failure(@Nullable PAssertionSite site) {
+      return new SuccessOrFailure(false, site, null);
+    }
+
+    @Override
+    public String toString() {
+      return "SuccessOrFailure{"
+          + (isSuccess ? "success" : "failure")
+          + ", error: "
+          + (assertionError() != null ? assertionError().toString() : "null")
+          + "}";
+    }
+  }
+  ////////////////////////////////////////////////////////////
+
+  /** Track the place where an assertion is defined.*/
+  protected static class PAssertionSite implements Serializable {
```

Comment:
Document why we needed to do this? Also, consider defining these static classes inside the SuccessOrFailure, and consider defining that in a separate PAssertSuccessOrFailure file (to keep the related code together)